### PR TITLE
feat(staker): New staker API (dashboard) [CHAIN-649]

### PR DESCRIPTION
**Pull request by DieMyst**

### CODE REVIEW

File: integration-tests-clients/package.json [hunk 1]
```diff
@@ -11,7 +11,9 @@
   "scripts": {
     "test-explorer": "vitest run dealExplorerClient.test.ts --no-file-parallelism",
     "test-matcher": "vitest run dealMatcherClient.test.ts --no-file-parallelism",
+    "test-staker": "vitest run stakerClient.test.ts --no-file-parallelism",
     "lint": "prettier src --check && eslint --cache \"./src/**/*.ts\"",
+    "fmt": "prettier 'src/**/*.ts' --w",
```
Comment (folex) [inline integration-tests-clients/package.json:16]: add check-fmt and add it to Makefile please. Also I guess gotta add it to `install-npms` in Makefile.

Would be very nice to unite it all into a single Workspace, yarn or pnpm. Some day.

File: integration-tests-clients/src/utils/events.ts [hunk 2]
```diff
@@ -19,6 +25,30 @@ type Contract<T> = {
   };
 };
 
+export function getEvent<T extends string, U extends Contract<T>>({
```
Comment (InversionSpaces) [inline integration-tests-clients/src/utils/events.ts:28]: Can `getEventValue` use this function?

File: integration-tests-clients/src/utils/events.ts [hunk 3]
```diff
@@ -10,6 +10,12 @@ type GetEventValueArgs<T extends string, U extends Contract<T>> = {
   value: string;
 };
 
+type GetEventArgs<T extends string, U extends Contract<T>> = {
```
Comment (InversionSpaces) [inline integration-tests-clients/src/utils/events.ts:13]: To keep types in sync one can be defined through another:
```typescript
type GetEventValueArgs<T extends string, U extends Contract<T>> = 
  GetEventArgs<T, U> & {value: string};
  ```

File: integration-tests-clients/src/utils/fixture.ts [hunk 4]
```diff
@@ -206,9 +207,15 @@ export async function depositCollateral(
...
-  });
+  const receipt = await sign(
+    capacityContract.depositCollateral,
+    capacityCommitmentIds,
+    {
+      value: valueToSendForCollateralAll,
+    },
+  );
+
+  return receipt.from as string;
```
Comment (InversionSpaces) [inline integration-tests-clients/src/utils/fixture.ts:218]: Why cast is needed?

Comment (InversionSpaces) [reply integration-tests-clients/src/utils/fixture.ts:218]: I also don't understand why return it all the way up. It should be known to caller code already.

File: ts-client/src/dealExplorerClient/serializers/stakerEncoders.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  const status = serializeStatus(filter.status);
+  if (status) {
+    filters.push({ status: status });
+  }
+
+  return {
+    and: filters,
+  };
```
Comment (InversionSpaces) [inline ts-client/src/dealExplorerClient/serializers/stakerEncoders.ts:78]: nit: can this be one object without nested `and`?

File: ts-client/src/dealExplorerClient/serializers/stakerEncoders.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+export function serializeOrderBy(
+  orderBy?: CapacityCommitmentsDashboardOrderBy,
```
Comment (InversionSpaces) [inline ts-client/src/dealExplorerClient/serializers/stakerEncoders.ts:82]: nit: Why make parameter optional, if function throws if it is not presented? If parameter is not optional, ts will complain on call site

File: ts-client/src/dealExplorerClient/serializers/stakerEncoders.ts [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  CapacityCommitment_OrderBy,
+  CapacityCommitmentStatus as CCS,
+} from "../indexerClient/generated.types.js";
+import type { CapacityCommitmentStatus } from "../types/schemes.js";
+import type {
+  CapacityCommitmentsDashboardFilter,
+  CapacityCommitmentsDashboardOrderBy,
+} from "../types/stakerFilters.js";
+
+export function unserializeStatus(
```
Comment (InversionSpaces) [inline ts-client/src/dealExplorerClient/serializers/stakerEncoders.ts:12]: nit: Common naming is **de**serialize

File: ts-client/src/dealExplorerClient/serializers/stakerEncoders.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+} from "../indexerClient/generated.types.js";
+import type { CapacityCommitmentStatus } from "../types/schemes.js";
+import type {
+  CapacityCommitmentsDashboardFilter,
+  CapacityCommitmentsDashboardOrderBy,
+} from "../types/stakerFilters.js";
+
+export function deserializeStatus(
+  ccs: CCS | null | undefined,
+): CapacityCommitmentStatus {
```
Comment (crystalbit) [inline ts-client/src/dealExplorerClient/serializers/stakerEncoders.ts:14]: shall it be `CCS | undefined`?

File: ts-client/src/dealExplorerClient/epochCalculator.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
+import { calculateTimestamp, calculateEpoch } from "./utils.js";
+
+export class EpochCalculator {
+  private readonly coreInitTimestamp: number;
+  private readonly coreEpochDuration: number;
+
+  constructor(coreInitTimestamp: number, coreEpochDuration: number) {
+    this.coreInitTimestamp = coreInitTimestamp;
+    this.coreEpochDuration = coreEpochDuration;
+  }
```
Comment (InversionSpaces) [inline ts-client/src/dealExplorerClient/epochCalculator.ts:10]:
```suggestion
  }
  
```

File: ts-client/src/dealExplorerClient/indexerClient/contractConstants.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
+import { IndexerClientABC } from "../../utils/indexerClient/indexerClientABC.js";
+import type { ContractsENV } from "../../client/config.js";
+import type { Sdk as ContractConstantsSdk } from "./queries/constants-query.generated.js";
+import { getSdk as getContractConstantsSdk } from "./queries/constants-query.generated.js";
```
Comment (InversionSpaces) [inline ts-client/src/dealExplorerClient/indexerClient/contractConstants.ts:4]:
```suggestion
import {
  type Sdk as ContractConstantsSdk,
  getSdk as getContractConstantsSdk,
} from "./queries/constants-query.generated.js";
```

File: ts-client/src/dealExplorerClient/indexerClient/contractConstants.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+export class ContractConstantsClient extends IndexerClientABC {
+  private contractConstantsClient: ContractConstantsSdk;
+  private contractConstants: ContractConstants | null;
+
+  constructor(network: ContractsENV) {
+    super(network);
+    this.contractConstantsClient = getContractConstantsSdk(this._graphqlClient);
+    this.contractConstants = null;
+  }
```
Comment (InversionSpaces) [inline ts-client/src/dealExplorerClient/indexerClient/contractConstants.ts:15]: This class is created just to retrieve contract constants. At all places where it is created instance of `IndexerClient` is created already. I don't understand why not just add contracts constants sdk to `IndexerClient` (as it is done with all other SDKs) and expose `ConstantsQuery` there. Then just deserialize query result to model type (as it is done with other query results). I don't understand why this class is needed.

File: ts-client/src/dealExplorerClient/serializers/stakerEncoders.ts (hunk 8 above)
Comment (InversionSpaces) [reply ts-client/src/dealExplorerClient/serializers/stakerEncoders.ts:14]: On the call site type is `CCS | null` and `null` is not subtype of `CCS | undefined`

File: ts-client/src/dealExplorerClient/indexerClient/indexerClient.ts [hunk 12]
```diff
@@ -33,21 +33,27 @@ import type {
...
 } from "./queries/capacity-commitments-query.generated.js";
-import type { Sdk as ContractConstantsSdk } from "./queries/constants-query.generated.js";
-import { getSdk as getContractConstantsSdk } from "./queries/constants-query.generated.js";
 import { getSdk as getCapacityCommitmentsSdk } from "./queries/capacity-commitments-query.generated.js";
 import { IndexerClientABC } from "../../utils/indexerClient/indexerClientABC.js";
 import type { ContractsENV } from "../../client/config.js";
 import type { CapacityCommitmentsQueryQueryVariables } from "./queries/capacity-commitments-query.generated.js";
 import type { PeerQueryQueryVariables } from "./queries/peers-query.generated.js";
 
+export interface ContractConstants {
```
Comment (InversionSpaces) [inline ts-client/src/dealExplorerClient/indexerClient/indexerClient.ts:42]: Does this interface belong here?

File: ts-client/src/dealExplorerClient/types/stakerFilters.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
+import type { CapacityCommitmentStatus } from "./schemes.js";
+
+export interface CapacityCommitmentsStakeFilters {
```
Comment (InversionSpaces) [inline ts-client/src/dealExplorerClient/types/stakerFilters.ts:3]: Where is this interface used?

File: integration-tests-clients/src/utils/events.ts (hunk 2 above)
Comment (InversionSpaces) [reply integration-tests-clients/src/utils/events.ts:26]: ping @DieMyst

File: integration-tests-clients/src/staker/stakerClient.test.ts [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+      // Create/update provider.
+      const setProviderInfoTx = await marketContract.setProviderInfo(
+        "ProviderWithActiveCC",
+        {
+          prefixes: "0x12345678",
+          hash: ethers.hexlify(ethers.randomBytes(32)),
+        },
+      );
+      await setProviderInfoTx.wait(WAIT_CONFIRMATIONS);
```
Comment (InversionSpaces) [inline integration-tests-clients/src/staker/stakerClient.test.ts:112]: Use `sign` from `utils.ts`

File: integration-tests-clients/src/staker/stakerClient.test.ts [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+
+      const data = result.data;
+      expect(data).toHaveLength(4);
+      expect(data[0].stakingReward).toBe(5);
+      expect(data[1].stakingReward).toBe(5);
+      expect(data[2].stakingReward).toBe(3);
+      expect(data[3].stakingReward).toBe(3);
+      expect(data[0].finishAt == data[1].finishAt).toBe(true);
+      expect(data[2].finishAt == data[3].finishAt).toBe(true);
+      expect(data[0].finishAt == data[2].finishAt).toBe(false);
```
Comment (InversionSpaces) [inline integration-tests-clients/src/staker/stakerClient.test.ts:159]: Can you refactor `3` and `5` to constants with meaningful names? 
Also add messages to `expect`s please

File: ts-client/src/dealExplorerClient/types/stakerFilters.ts (hunk 13 above)
Comment (DieMyst) [reply ts-client/src/dealExplorerClient/types/stakerFilters.ts:3]: it will be used on another page. It is not used now

**Review by InversionSpaces — APPROVED** [review]

**Pull request merged**